diff --git a/test/lightning/version_control/github_client_test.exs b/test/lightning/version_control/github_client_test.exs
index 1b63a5df5..9aaddb969 100644
--- a/test/lightning/version_control/github_client_test.exs
+++ b/test/lightning/version_control/github_client_test.exs
@@ -19,6 +19,9 @@ defmodule Lightning.VersionControl.GithubClientTest do
           "https://api.github.com/app/installations/some-id/access_tokens" ->
             %Tesla.Env{status: 200, body: %{"token" => "some-token"}}
 
+          "https://api.github.com/app/installations/invalid/access_tokens" ->
+            %Tesla.Env{status: 401, body: %{"message" => "Bad credentials"}}
+
           "https://api.github.com/installation/repositories" ->
             %Tesla.Env{
               status: 200,
@@ -44,6 +47,11 @@ defmodule Lightning.VersionControl.GithubClientTest do
     test "client can fetch repo branches" do
       p_repo = insert(:project_repo)
 
+      assert {:ok, ["master"]} =
+               VersionControl.fetch_repo_branches(p_repo.project_id, p_repo.repo)
+
+      p_repo = insert(:project_repo, github_installation_id: "invalid")
+
       assert {:ok, ["master"]} =
                VersionControl.fetch_repo_branches(p_repo.project_id, p_repo.repo)
     end
